File: R/metabolite_set.R
# =============================================================================
# metabolite_set : an S3 class holding a wide metabolomics table plus the
# explicitly-resolved set of metabolite (feature) columns.
#
# Identification happens ONCE, at construction, and is validated there.
# Every downstream method reads x$metabolite_cols instead of re-guessing.
# =============================================================================

#' Resolve which columns are metabolites
#'
#' Three mutually exclusive strategies, in priority order:
#' \describe{
#'   \item{\code{metabolites}}{An explicit character vector of column names.}
#'   \item{\code{pattern}}{A regex matched against \code{names(data)}
#'     (e.g. \code{"^HMDB"}).}
#'   \item{\code{id_cols}}{Names of metadata columns; every \emph{other} numeric
#'     column is treated as a metabolite. Usually the most robust choice.}
#' }
#'
#' @param data A data.frame.
#' @param metabolites,pattern,id_cols See Details; supply exactly one.
#' @return A character vector of column names.
#' @keywords internal
#' @noRd
resolve_metabolite_cols <- function(data, metabolites = NULL,
                                    pattern = NULL, id_cols = NULL) {
  supplied <- c(!is.null(metabolites), !is.null(pattern), !is.null(id_cols))
  if (sum(supplied) == 0)
    stop("Specify exactly one of `metabolites`, `pattern`, or `id_cols`.",
         call. = FALSE)
  if (sum(supplied) > 1)
    stop("Specify only one of `metabolites`, `pattern`, or `id_cols`.",
         call. = FALSE)
  
  if (!is.null(metabolites)) {
    missing <- setdiff(metabolites, names(data))
    if (length(missing))
      stop("Metabolite columns not found in data: ",
           paste(missing, collapse = ", "), call. = FALSE)
    cols <- metabolites
  } else if (!is.null(pattern)) {
    cols <- grep(pattern, names(data), value = TRUE)
  } else {
    missing <- setdiff(id_cols, names(data))
    if (length(missing))
      stop("`id_cols` not found in data: ",
           paste(missing, collapse = ", "), call. = FALSE)
    candidate <- setdiff(names(data), id_cols)
    cols <- candidate[vapply(data[candidate], is.numeric, logical(1))]
  }
  
  if (length(cols) == 0)
    stop("No metabolite columns matched the given specification.", call. = FALSE)
  cols
}

#' Bare constructor (internal, no validation)
#'
#' @param data A data.frame.
#' @param metabolite_cols Character vector of resolved column names.
#' @return An object of class \code{"metabolite_set"}.
#' @keywords internal
#' @noRd
new_metabolite_set <- function(data, metabolite_cols) {
  stopifnot(is.data.frame(data), is.character(metabolite_cols))
  structure(
    list(data = data, metabolite_cols = metabolite_cols),
    class = "metabolite_set"
  )
}

#' Validate a metabolite_set's invariants
#'
#' @param x A candidate \code{metabolite_set}.
#' @return \code{x}, invisibly, if valid; otherwise an error.
#' @keywords internal
#' @noRd
validate_metabolite_set <- function(x) {
  cols <- x$metabolite_cols
  
  missing <- setdiff(cols, names(x$data))
  if (length(missing))
    stop("metabolite_cols refer to absent columns: ",
         paste(missing, collapse = ", "), call. = FALSE)
  
  non_numeric <- cols[!vapply(x$data[cols], is.numeric, logical(1))]
  if (length(non_numeric))
    stop("Metabolite columns must be numeric; these are not: ",
         paste(non_numeric, collapse = ", "), call. = FALSE)
  
  x
}

#' Create a metabolite_set
#'
#' User-facing constructor. Resolves the metabolite columns from one of three
#' strategies, then validates the result.
#'
#' @param data A data.frame in wide format (one row per sample, one column per
#'   metabolite, plus any metadata columns).
#' @param metabolites Optional explicit character vector of metabolite columns.
#' @param pattern Optional regex matched against column names.
#' @param id_cols Optional names of metadata columns; the remaining numeric
#'   columns are taken to be metabolites.
#' @return An object of class \code{"metabolite_set"}.
#' @export
#' @examples
#' df <- data.frame(
#'   sample_id = c("a", "b", "c"),
#'   group     = c("x", "y", "x"),
#'   glucose   = c(0, 2.1, 3.4),
#'   lactate   = c(1.2, 0, 0.9)
#' )
#' ms <- metabolite_set(df, id_cols = c("sample_id", "group"))
#' ms
metabolite_set <- function(data, metabolites = NULL,
                           pattern = NULL, id_cols = NULL) {
  cols <- resolve_metabolite_cols(data, metabolites, pattern, id_cols)
  validate_metabolite_set(new_metabolite_set(data, cols))
}

#' @export
format.metabolite_set <- function(x, ...) {
  eg <- utils::head(x$metabolite_cols, 3L)
  more <- length(x$metabolite_cols) - length(eg)
  eg_str <- paste(eg, collapse = ", ")
  if (more > 0) eg_str <- paste0(eg_str, ", ... (", more, " more)")
  c(
    "<metabolite_set>",
    paste0("  samples:     ", nrow(x$data)),
    paste0("  metabolites: ", length(x$metabolite_cols)),
    paste0("  columns:     ", eg_str)
  )
}

#' @export
print.metabolite_set <- function(x, ...) {
  cat(format(x, ...), sep = "\n")
  invisible(x)
}

File: Python/metabolite_set.py
"""
metabolite_set : a small class holding a wide metabolomics table plus the
explicitly-resolved set of metabolite (feature) columns.

Port of met_cor.R / metabolite_set.R. Identification happens ONCE, at
construction, and is validated there. Every downstream function reads
`ms.metabolite_cols` instead of re-guessing.

The R original used S3 (an object + generics dispatching on class). Python's
natural equivalent is a plain class with methods, so the S3 `UseMethod`
plumbing simply disappears.
"""

from __future__ import annotations

import re
from typing import Iterable, Optional, Sequence

import polars as pl
import tidypolars as tp


def _as_tibble(data) -> tp.tibble:
    """Accept a tidypolars tibble, polars/pandas DataFrame, or dict."""
    if isinstance(data, tp.tibble):
        return data
    if isinstance(data, pl.DataFrame):
        return tp.from_polars(data)
    # pandas / dict / anything polars can construct from
    try:
        import pandas as pd  # noqa
        if isinstance(data, pd.DataFrame):
            return tp.from_pandas(data)
    except ImportError:
        pass
    return tp.from_polars(pl.DataFrame(data))


def resolve_metabolite_cols(
    data: tp.tibble,
    metabolites: Optional[Sequence[str]] = None,
    pattern: Optional[str] = None,
    id_cols: Optional[Sequence[str]] = None,
) -> list[str]:
    """
    Resolve which columns are metabolites via exactly one of three strategies:

    - `metabolites`: an explicit list of column names.
    - `pattern`:     a regex matched against the column names (e.g. r"^HMDB").
    - `id_cols`:     names of metadata columns; every *other* numeric column is
                     treated as a metabolite. Usually the most robust choice.
    """
    supplied = [metabolites is not None, pattern is not None, id_cols is not None]
    if sum(supplied) == 0:
        raise ValueError("Specify exactly one of `metabolites`, `pattern`, or `id_cols`.")
    if sum(supplied) > 1:
        raise ValueError("Specify only one of `metabolites`, `pattern`, or `id_cols`.")

    names = list(data.names)
    schema = dict(zip(data.names, data.dtypes))  # tibble blocks .schema

    if metabolites is not None:
        missing = [m for m in metabolites if m not in names]
        if missing:
            raise ValueError("Metabolite columns not found in data: " + ", ".join(missing))
        cols = list(metabolites)
    elif pattern is not None:
        cols = [n for n in names if re.search(pattern, n)]
    else:  # id_cols
        missing = [c for c in id_cols if c not in names]
        if missing:
            raise ValueError("`id_cols` not found in data: " + ", ".join(missing))
        candidate = [n for n in names if n not in set(id_cols)]
        cols = [c for c in candidate if schema[c].is_numeric()]

    if len(cols) == 0:
        raise ValueError("No metabolite columns matched the given specification.")
    return cols


class metabolite_set:
    """
    A wide metabolomics table + the resolved metabolite column names.

    Construct with exactly one of `metabolites`, `pattern`, or `id_cols`.
    `.data` is a tidypolars tibble; `.metabolite_cols` is a list[str].
    """

    def __init__(
        self,
        data,
        metabolites: Optional[Sequence[str]] = None,
        pattern: Optional[str] = None,
        id_cols: Optional[Sequence[str]] = None,
    ):
        tib = _as_tibble(data)
        cols = resolve_metabolite_cols(tib, metabolites, pattern, id_cols)
        self.data: tp.tibble = tib
        self.metabolite_cols: list[str] = cols
        self._validate()

    def _validate(self) -> "metabolite_set":
        names = set(self.data.names)
        missing = [c for c in self.metabolite_cols if c not in names]
        if missing:
            raise ValueError("metabolite_cols refer to absent columns: " + ", ".join(missing))
        schema = dict(zip(self.data.names, self.data.dtypes))
        non_numeric = [c for c in self.metabolite_cols if not schema[c].is_numeric()]
        if non_numeric:
            raise ValueError(
                "Metabolite columns must be numeric; these are not: " + ", ".join(non_numeric)
            )
        return self

    # --- printing: mirrors format.metabolite_set / print.metabolite_set ------
    def __repr__(self) -> str:
        eg = self.metabolite_cols[:3]
        more = len(self.metabolite_cols) - len(eg)
        eg_str = ", ".join(eg)
        if more > 0:
            eg_str += f", ... ({more} more)"
        return (
            "<metabolite_set>\n"
            f"  samples:     {self.data.nrow}\n"
            f"  metabolites: {len(self.metabolite_cols)}\n"
            f"  columns:     {eg_str}"
        )
